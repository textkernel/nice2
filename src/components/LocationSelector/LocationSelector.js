import React from 'react';
import PropTypes from 'prop-types';
import bem from 'bem';
import { LoadScriptNext } from '@react-google-maps/api';
import { Text, Modal, Map, FieldWrapper, Button, LocationCard } from '../../index';
import styles from './LocationSelector.scss';

const { block, elem } = bem({
    name: 'LocationSelector',
    classnames: styles,
});

function LocationSelector(props) {
    const {
        /** FieldWrapper props */
        clearLabel,
        onRemoveAllLocations,

        /** LocationCard props */
        minRadius,
        maxRadius,
        radiusDefaultValue,
        selectedLocations,
        onAddLocation,
        onUpdateLocation,
        onRemoveLocation,
        doneLabel,

        /** LocationAutocomplete props */
        country,
        placeTypes,
        autocompletePlaceholder,
        noSuggestionsPlaceholder,
        showCountryInSuggestions,
        onLocationAutocompleteError,
    } = props;

    const [isOpen, setIsOpen] = React.useState(false);

    function handleOpenModal() {
        if (!isOpen) {
            setIsOpen(true);
        }
    }

    function handleCloseModal() {
        if (isOpen) {
            setIsOpen(false);
        }
    }

    /**
     * Fetch additional information for the selected place and
     * add it along with passed location object to the selectedLocations array
     */
    function handleAddLocation(location) {
        // const placeInfo = fetchPlaceInfo(location.id);
        // const locationToAdd = {
        //     ...location,
        //     center: placeInfo.coordinates,
        //     radius: radiusDefaultValue,
        //     structuredFormatting: placeInfo.structuredFormatting,
        // };
        // onAddLocation(locationToAdd);
    }

    return (
        <div {...block(props)}>
            <FieldWrapper
                onClick={handleOpenModal}
                clearLabel={clearLabel}
                onClear={onRemoveAllLocations}
            >
                <Text placeholder="Some placeholder" {...elem('mainTextInput', props)}>
                    Some placeholder
                </Text>
            </FieldWrapper>
            <Modal {...elem('modal', props)} isOpen={isOpen} onRequestClose={handleCloseModal}>
                <LoadScriptNext>
                    <div {...elem('autocomplete', props)}>
                        <div {...elem('input', props)} />
                        {/* <LocationAutocomplete
                            inputPlaceholder={autocompletePlaceholder}
                            noSuggestionsPlaceholder={noSuggestionsPlaceholder}
                            onSelectionChange={handleAddLocation}
                            country={country}
                            placeTypes={placeTypes}
                            showCountryInSuggestions={showCountryInSuggestions}
                            onError={onLocationAutocompleteError}
                        /> */}
                        <Button
                            {...elem('button', props)}
                            onClick={handleCloseModal}
                            context="brand"
                        >
                            {doneLabel}
                        </Button>
                    </div>
                    <div {...elem('locationsWrapper', props)}>
                        <ul {...elem('locationCardsContainer', props)}>
                            {selectedLocations.map(location => (
                                <LocationCard
                                    {...elem('locationCard', props)}
                                    As="li"
                                    locationId={location.id}
                                    locationTitle={location.description}
                                    distanceRadius={location.radius}
                                    sliderLabel={location.sliderLabel}
                                    minRadius={minRadius}
                                    maxRadius={maxRadius}
                                    onRadiusChange={onUpdateLocation}
                                    onDelete={onRemoveLocation}
                                />
                            ))}
                        </ul>
                        <Map />
                    </div>
                </LoadScriptNext>
            </Modal>
        </div>
    );
}

LocationSelector.displayName = 'LocationSelector';

LocationSelector.propTypes = {
    /** Google api key */
    apiKey: PropTypes.string.isRequired,
    /** stores an array of selected location objects */
    selectedLocations: PropTypes.arrayOf(
        PropTypes.shape({
            id: PropTypes.string.isRequired,
            description: PropTypes.string.isRequired,
            structuredFormatting: PropTypes.string.isRequired,
            center: PropTypes.shape({
                lng: PropTypes.number.isRequired,
                lan: PropTypes.number.isRequired,
            }).isRequired,
            radius: PropTypes.number.isRequired,
        })
    ).isRequired,
    /** country where search can take place */
    country: PropTypes.string.isRequired,
    /** language in which suggestions should be displayed */
    language: PropTypes.string.isRequired,
    /** default radius value */
    radiusDefaultValue: PropTypes.number,
    /** radius measurement unit */
    radiusUnits: PropTypes.oneOf(['km', 'mi']).isRequired,
    /** radius label suffix (e.g. km, miles or other language) */
    radiusUnitDisplayText: PropTypes.string.isRequired,
    /** min radius value of the slider component */
    minRadius: PropTypes.number.isRequired,
    /** max radius value of the slider component */
    maxRadius: PropTypes.number.isRequired,
    /** radius step value of the slider component */
    radiusStep: PropTypes.number.isRequired,
    /** map center coordinates if there's no selected locations */
    mapCenter: PropTypes.shape({
        lng: PropTypes.number.isRequired,
        lat: PropTypes.number.isRequired,
    }),
    /** map localize */
    mapRegion: PropTypes.string,
    /** map default zoom value */
    mapZoom: PropTypes.number,
    /** placeholder for LocationAutocomplete field */
    autocompletePlaceholder: PropTypes.string.isRequired,
    /** placeholder for empty LocationAutocomplete list */
    noSuggestionsPlaceholder: PropTypes.string.isRequired,
    /**
     * type of locations that should be searched for.
     * For details see: https://developers.google.com/maps/documentation/javascript/reference/places-autocomplete-service#AutocompletionRequest.types
     */
    // placeTypes: LocationAutocomplete.propTypes.placeTypes,
    /** function to be executed if error occurs while fetching suggestions */
    onLocationAutocompleteError: PropTypes.func,
    /** string to be displayed in FieldWrapper when the modal is closed, but locations are selected */
    selectionPlaceholder: PropTypes.string.isRequired,
    /** placeholder for FieldWrapper when there's no locations selected */
    mainPlaceholder: PropTypes.string.isRequired,
    /** label for the Done button */
    doneLabel: PropTypes.string.isRequired,
    /** show country name in autocomplete suggestions */
    showCountryInSuggestions: PropTypes.bool,
    /** label to be used on the clear all button */
    clearLabel: PropTypes.string.isRequired,
    /** function with a location object as an argument to be added to the selectedLocations array */
    onAddLocation: PropTypes.func.isRequired,
    /** function with a location details as an argument to be changed */
    onUpdateLocation: PropTypes.func.isRequired,
    /** function with a locationId as an argument to be removed */
    onRemoveLocation: PropTypes.func.isRequired,
    /** callback function for the Clear button click */
    onRemoveAllLocations: PropTypes.func.isRequired,
    /** callback function for closed modal */
    onBlur: PropTypes.func,
};

LocationSelector.defaultProps = {
    // placeTypes: LocationAutocomplete.propTypes.placeTypes,
    radiusDefaultValue: 1,
    showCountryInSuggestions: true,
    onBlur: () => null,
};

export default LocationSelector;
