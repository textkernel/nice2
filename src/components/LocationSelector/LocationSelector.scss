$modal-width: 75%;
$done-button-width: 150px;
$location-card-width: 300px;

.LocationSelector {
    &__mainTextInput {
        border: 0;
        margin: var(--spacing-normal);
        font-size: var(--font-size-normal);
        color: var(--color-muted);
    }

    &__modal {
        display: flex;
        flex-direction: column;
        width: $modal-width;
    }

    &__autocomplete {
        display: flex;
        box-sizing: border-box;
        width: 100%;
        padding: var(--spacing-4x);
    }

    &__input {
        width: 100%;
        border: 1px double black;
    }

    &__button {
        width: $done-button-width;
        margin-left: var(--spacing-4x);
    }

    &__locationsWrapper {
        display: flex;
        height: 100%;
    }

    &__locationCardsContainer {
        padding: 0;
        margin: 0;
        width: $location-card-width;
        min-width: $location-card-width;
        padding-right: calc(var(--spacing-4x) - 1px);
        overflow: scroll;
    }

    &__locationCard:not(:last-child) {
        margin-bottom: var(--spacing-normal);
    }
}
