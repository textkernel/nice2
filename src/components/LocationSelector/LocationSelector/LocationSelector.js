import React from 'react';
import PropTypes from 'prop-types';
import bem from 'bem';
import { Map, Button, LocationCard, LocationAutocomplete } from '../../../index';
import styles from './LocationSelector.scss';

const { elem } = bem({
    name: 'LocationSelector',
    classnames: styles,
});

const LocationSelector = props => {
    const {
        /** FieldWrapper props */
        inputPlaceholder,

        /** LocationCard props */
        minRadius,
        maxRadius,
        radiusStep,
        renderRadiusLabel,
        onRemoveLocation,
        doneLabel,

        /** LocationAutocomplete props */
        country,
        placeTypes,
        noSuggestionsPlaceholder,
        showCountryInSuggestions,
        onLocationAutocompleteError,

        /** Internal use */
        onUpdateLocation,
        selectedLocations,
        getMarkers,
        onAddLocation,
        onCloseModal,
    } = props;

    return (
        <>
            <div {...elem('inputLine', props)}>
                <LocationAutocomplete
                    {...elem('searchField', props)}
                    isFocused
                    inputPlaceholder={inputPlaceholder}
                    noSuggestionsPlaceholder={noSuggestionsPlaceholder}
                    onSelectionChange={onAddLocation}
                    country={country}
                    placeTypes={placeTypes}
                    showCountryInSuggestions={showCountryInSuggestions}
                    onError={onLocationAutocompleteError}
                    hidePoweredByGoogleLogo
                />
                <Button {...elem('button', props)} onClick={onCloseModal} context="brand">
                    {doneLabel}
                </Button>
            </div>
            <div {...elem('locationsWrapper', props)}>
                {!!selectedLocations.length && (
                    <ul {...elem('locationCardsContainer', props)}>
                        {selectedLocations.map(location => (
                            <LocationCard
                                {...elem('locationCard', props)}
                                As="li"
                                key={location.id}
                                locationId={location.id}
                                locationTitle={location.description}
                                distanceRadius={location.radius}
                                sliderLabel={renderRadiusLabel(location.radius)}
                                minRadius={minRadius}
                                maxRadius={maxRadius}
                                radiusStep={radiusStep}
                                onRadiusChange={onUpdateLocation}
                                onDelete={onRemoveLocation}
                            />
                        ))}
                    </ul>
                )}
                <Map defaultArea={{ address: country }} markers={getMarkers()} />
            </div>
        </>
    );
};

LocationSelector.displayName = 'LocationSelector';

LocationSelector.propTypes = {
    /** stores an array of selected location objects */
    selectedLocations: PropTypes.arrayOf(
        PropTypes.shape({
            id: PropTypes.string.isRequired,
            description: PropTypes.string.isRequired,
            center: PropTypes.shape({
                lng: PropTypes.number.isRequired,
                lat: PropTypes.number.isRequired,
            }).isRequired,
            radius: PropTypes.number.isRequired,
        })
    ).isRequired,
    /** radius label renderer e.g. radius => `+ ${radius} km` */
    renderRadiusLabel: PropTypes.func.isRequired,
    /** min radius value of the slider component */
    minRadius: PropTypes.number.isRequired,
    /** max radius value of the slider component */
    maxRadius: PropTypes.number.isRequired,
    /** radius step value of the slider component */
    radiusStep: PropTypes.number.isRequired,
    /** country where search can take place */
    country: PropTypes.string.isRequired,
    /**
     * type of locations that should be searched for.
     * For details see: https://developers.google.com/maps/documentation/javascript/reference/places-autocomplete-service#AutocompletionRequest.types
     */
    placeTypes: PropTypes.arrayOf(PropTypes.string).isRequired,
    /** show country name in autocomplete suggestions */
    showCountryInSuggestions: PropTypes.bool.isRequired,
    /** placeholder for both main field and autocomplete field in modal */
    inputPlaceholder: PropTypes.string.isRequired,
    /** placeholder for empty LocationAutocomplete list */
    noSuggestionsPlaceholder: PropTypes.string.isRequired,
    /** function to be executed if error occurs while fetching suggestions */
    onLocationAutocompleteError: PropTypes.func.isRequired,
    /** label for the Done button */
    doneLabel: PropTypes.string.isRequired,
    /** function called with location object as an argument when it is selected from the suggestions */
    onAddLocation: PropTypes.func.isRequired,
    /** function called with a location details as an argument to be changed */
    onUpdateLocation: PropTypes.func.isRequired,
    /** function with a locationId as an argument to be removed */
    onRemoveLocation: PropTypes.func.isRequired,
    getMarkers: PropTypes.func.isRequired,
};

export default LocationSelector;
