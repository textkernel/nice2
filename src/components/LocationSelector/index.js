export { LocationSelectorDialog } from './LocationSelectorDialog';
export { LocationSelectorDialogWithGoogleLoader } from './LocationSelectorDialogWithGoogleLoader';
export { LocationSelector } from './LocationSelector';
export { LocationSelectorLocation, LocationSelectorAddressComponent } from './utils';
