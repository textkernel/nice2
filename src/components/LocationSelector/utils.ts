const METER_TO_KM = 1000;
const METER_TO_MI = 1609.34;

export type Location = {
    id: string;
    description: string;
    center: {
        lng: number;
        lat: number;
    };
    radius: number;
};

export const getRadiusInMeters = (radius, radiusUnits) => {
    if (radiusUnits === 'km') {
        return radius * METER_TO_KM;
    }
    return radius * METER_TO_MI;
};

export const findCenter = (geocoder, placeId) =>
    new Promise(function geocoderPromise(resolve, reject) {
        geocoder.geocode({ placeId }, (results, status) => {
            if (status === 'OK') {
                if (results[0]) {
                    resolve(results[0].geometry.location);
                } else {
                    reject(new Error(`No results found when searching for placeId ${placeId}`));
                }
            } else {
                reject(new Error(`Geocoder failed due to: ${status}`));
            }
        });
    });

export const getAddressComponents = (geocoder, location) =>
    new Promise(function geocoderPromise(resolve, reject) {
        geocoder.geocode({ location }, (results, status) => {
            if (status === 'OK') {
                if (results[0]) {
                    resolve(results[0].address_components);
                } else {
                    reject(
                        new Error(
                            `No results found when searching for coordinates ${location.lat}, ${location.lng}`
                        )
                    );
                }
            } else {
                reject(new Error(`Geocoder failed due to: ${status}`));
            }
        });
    });
