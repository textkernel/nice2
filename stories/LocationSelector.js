import React from 'react';
import { storiesOf } from '@storybook/react';
import { number, text, select, boolean, withKnobs } from '@storybook/addon-knobs';
import LocationSelector from '../src/components/LocationSelector';
import ensureApiKey from './utils/ensureApiKey';
import withStore from '../src/packages/storybook/withStore';

const selectedLocations = [
    {
        locationId: 'ajdo-219a-j19v-0491',
        description: 'London',
        radius: 42,
        sliderLabel: '42km',
    },
    {
        locationId: 'ajdo-219a-j19v-0492',
        description: 'Berlin',
        radius: 20,
        sliderLabel: '20km',
    },
    {
        locationId: 'ajdo-219a-j19v-04925',
        description: 'Berlin',
        radius: 20,
        sliderLabel: '20km',
    },
    {
        locationId: 'ajdo-219a-j19v-0494',
        description: 'Berlin',
        radius: 20,
        sliderLabel: '20km',
    },
    {
        locationId: 'ajdo-219a-j19v-0492231',
        description: 'Berlin',
        radius: 20,
        sliderLabel: '20km',
    },
];

storiesOf('Organisms|LocationSelector', module)
    .addDecorator(withKnobs)
    .addParameters(
        withStore({
            selectedLocations: [],
        })
    )
    .add('Basic component', () => {
        const apiKey = ensureApiKey();

        return (
            <LocationSelector
                apiKey={apiKey}
                selectedLocations={selectedLocations}
                country={text('country', 'NL')}
                language={text('Language', 'EN')}
                radiusUnits="km"
                renderRadiusLabel={r => `+ ${r} km`}
                minRadius={number('Min radius', 1)}
                maxRadius={number('Max radius', 100)}
                radiusStep={number('Radius steps', 1)}
                autocompletePlaceholder="autocompletePlaceholder"
                noSuggestionsPlaceholder="noSuggestionsPlaceholder"
                selectionPlaceholder="selectionPlaceholder"
                mainPlaceholder="mainPlaceholder"
                doneLabel={text('Label for Done button', 'Done')}
                clearLabel={text('Label for clear button', 'Clear')}
                contentLabel={text('Placeholder for input field', 'Location selector')}
                onAddLocation={location => console.log('onAddLocation was called with:', location)}
                onUpdateLocation={() => console.log('onUpdateLocation was called')}
                onRemoveLocation={() => console.log('onRemoveLocation was called')}
                onRemoveAllLocations={() => console.log('onRemoveAllLocations was called')}
            />
        );
    })
    .add('Example implementation', ({ parameters }) => {
        const apiKey = ensureApiKey();
        const store = parameters.getStore();

        const handleAddLocation = location => {
            console.log('onAddLocation was called with:', location);
            store.set({ selectedLocations: [...store.get('selectedLocations'), location] });
        };

        const handleUpdateLocation = (id, radius) => {
            console.log('onUpdateLocation was called', id, radius);
            const newSelection = store.get('selectedLocations').map(loc => {
                if (loc.id === id) {
                    loc.radius = radius; // eslint-disable-line no-param-reassign
                }
                return loc;
            });
            store.set({ selectedLocations: newSelection });
        };

        const handleRemoveLocation = id => {
            console.log('onRemoveLocation was called', id);
            store.set({
                selectedLocations: store.get('selectedLocations').filter(l => l.id !== id),
            });
        };

        const handleRemoveAllLocations = () => {
            console.log('onRemoveAllLocations was called');
            store.set({ selectedLocations: [] });
        };

        const getPlaceholder = () => {
            const selection = store.get('selectedLocations');
            return selection.length ? selection[0].structured_formatting.main_text : null;
        };

        return (
            <LocationSelector
                apiKey={apiKey}
                selectedLocations={store.get('selectedLocations')}
                country={text('country', 'NL')}
                language={text('Language', 'EN')}
                radiusUnits={select('Radius units', ['km', 'mi'], 'km')}
                renderRadiusLabel={r => `+ ${r} km`}
                minRadius={number('Min radius', 1)}
                maxRadius={number('Max radius', 100)}
                radiusDefaultValue={number('Radius default value', 5)}
                radiusStep={number('Radius steps', 1)}
                placeTypes="city"
                showCountryInSuggestions={boolean('Show country in suggestions', true)}
                modalContentLabel={text(
                    'Modal content title for screen readers',
                    'Location selection'
                )}
                inputPlaceholder={text('Input placeholder', 'Location...')}
                noSuggestionsPlaceholder="noSuggestionsPlaceholder"
                selectionPlaceholder={getPlaceholder()}
                doneLabel={text('Label for Done button', 'Done')}
                clearLabel={text('Label for clear button', 'Clear')}
                onAddLocation={handleAddLocation}
                onUpdateLocation={handleUpdateLocation}
                onRemoveLocation={handleRemoveLocation}
                onRemoveAllLocations={handleRemoveAllLocations}
                onBlur={() => {
                    console.log('onBlur was called');
                }}
            />
        );
    });
