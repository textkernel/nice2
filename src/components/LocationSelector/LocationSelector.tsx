import * as React from 'react';
import { FaMapMarkerAlt } from 'react-icons/fa';
import { bem } from '../../utils';
import { Modal } from '../Modal';
import { FieldWrapper } from '../FieldWrapper';
import { LocationSelectorDialogWithGoogleLoader } from './LocationSelectorDialogWithGoogleLoader';
import { Location, findCenter, getRadiusInMeters } from './utils';
import { TAB_KEY, ESCAPE_KEY } from '../../constants';
import styles from './LocationSelector.scss';

const { block, elem } = bem('LocationSelector', styles);

interface Props {
    /** define id for input element */
    id?: string;
    /** Google api key */
    apiKey: string;
    /** language in which suggestions should be displayed */
    language: string;
    /** Regional setting for the map. By default Google uses US.
     * For details see: https://developers.google.com/maps/documentation/javascript/localization#Region
     */
    region?: string;
    /** other props to pass to the google loader. For details see: https://react-google-maps-api-docs.netlify.com/#loadscriptnext */
    additionalGoogleProps?: object; // eslint-disable-line react/forbid-prop-types
    /** stores an array of selected location objects */
    selectedLocations: Location[];
    /** defines if selector has an option to control the radius for a marker */
    hasRadius?: boolean;
    /** default radius value */
    radiusDefaultValue?: number;
    /** radius measurement unit */
    radiusUnits: 'km' | 'mi';
    /** radius label renderer e.g. radius => `+ ${radius} km` */
    renderRadiusLabel: (string) => string;
    /** min radius value of the slider component */
    minRadius?: number;
    /** max radius value of the slider component */
    maxRadius?: number;
    /** radius step value of the slider component */
    radiusStep?: number;
    /** defines if selector has a list of locations cards to render */
    withoutLocationCards?: boolean;
    /** country where search can take place */
    country: string;
    /** address to make initial map centering more specific */
    initialMapAddress?: string;
    /**
     * type of locations that should be searched for.
     * For details see: https://developers.google.com/maps/documentation/javascript/reference/places-autocomplete-service#AutocompletionRequest.types
     */
    placeTypes?: string[];
    /** show country name in autocomplete suggestions */
    showCountryInSuggestions?: boolean;
    /** A title for the modal that will be used by screenreaders */
    modalContentLabel: string;
    /** placeholder for both main field and autocomplete field in modal */
    inputPlaceholder: string;
    /** placeholder for empty LocationAutocomplete list */
    noSuggestionsPlaceholder: string;
    /** function to be executed if error occurs while fetching suggestions */
    onLocationAutocompleteError: () => void;
    /** string to be displayed in FieldWrapper when the modal is closed, but locations are selected */
    selectionPlaceholder?: string;
    /** label for the Done button */
    doneLabel: string;
    /** label to be used on the clear all button */
    clearLabel: string;
    /** function called with location object as an argument when it is selected from the suggestions */
    onAddLocation: (Location) => void;
    /** function called with a location details as an argument to be changed */
    onUpdateLocation: (Location) => void;
    /** function with a locationId as an argument to be removed */
    onRemoveLocation: (Location) => void;
    /** callback function for the Clear button click */
    onRemoveAllLocations: () => void;
    /** callback function for closed modal */
    onBlur: () => void;
}

export const LocationSelector: React.FC<Props> = (props) => {
    const {
        /** Google props */
        apiKey,
        language,
        region,
        additionalGoogleProps,

        /** FieldWrapper props */
        id,
        clearLabel,
        onRemoveAllLocations,
        inputPlaceholder,
        selectionPlaceholder,

        /** Modal props */
        modalContentLabel,

        /** LocationCard props */
        hasRadius,
        minRadius,
        maxRadius,
        radiusStep,
        renderRadiusLabel,
        doneLabel,

        /** LocationAutocomplete props */
        country,
        initialMapAddress,
        placeTypes,
        noSuggestionsPlaceholder,
        showCountryInSuggestions,
        onLocationAutocompleteError,

        /** Internal use */
        selectedLocations,
        radiusDefaultValue,
        radiusUnits,
        withoutLocationCards,
        onAddLocation,
        onUpdateLocation,
        onRemoveLocation,
        onBlur,
        ...rest
    } = props;

    const [isOpen, setIsOpen] = React.useState(false);
    const [isWrapperFocused, setIsWrapperFocused] = React.useState(false);
    const buttonRef = React.useRef<HTMLButtonElement>();

    const hasLocationsSelected = selectedLocations && selectedLocations.length > 0;

    function handleOpenModal() {
        if (!isOpen && !isWrapperFocused) {
            buttonRef.current?.focus();
            setIsOpen(true);
        }
        setIsWrapperFocused(true);
    }

    function handleCloseModal() {
        if (isOpen) {
            setIsOpen(false);
            onBlur();
        }
    }

    function handleButtonKeyPress(e: React.KeyboardEvent<HTMLButtonElement>) {
        if (e.key === ESCAPE_KEY) {
            buttonRef.current?.blur();
        } else if (!(e.key === TAB_KEY || e.key === ESCAPE_KEY)) {
            setIsOpen(true);
        }
    }

    function handleButtonBlur() {
        if (!isOpen) {
            setIsWrapperFocused(false);
        }
    }

    /**
     * Fetch additional information for the selected place and
     * add it along with passed location object to the selectedLocations array
     * if this location was not selected yet
     */
    function handleAddLocation(location) {
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        const { Geocoder } = (window as any).google.maps;
        const geocoder = new Geocoder();

        return findCenter(geocoder, location.place_id)
            .then((center: any) => {
                const locationToAdd = {
                    ...location,
                    center: {
                        lng: typeof center.lng === 'function' ? center.lng() : center.lng,
                        lat: typeof center.lat === 'function' ? center.lat() : center.lat,
                    },
                    radius: hasRadius ? radiusDefaultValue : 0,
                };

                const isLocationSelected = selectedLocations
                    .map((item) => item.id)
                    .includes(location.id);

                if (!isLocationSelected) {
                    onAddLocation(locationToAdd);
                }
            })
            .catch(/* TODO: add error handling */);
    }

    function getMarkers() {
        return selectedLocations.map((location) => ({
            description: location.description,
            center: location.center,
            radius: hasRadius ? getRadiusInMeters(location.radius, radiusUnits) : 0,
        }));
    }

    return (
        <div {...rest} {...block(props)}>
            <FieldWrapper
                {...elem('mainTextButtonWrapper', props)}
                isFocused={isWrapperFocused}
                showClearButton={hasLocationsSelected}
                clearLabel={clearLabel}
                onClick={handleOpenModal}
                onClear={onRemoveAllLocations}
            >
                <FaMapMarkerAlt {...elem('icon', props)} />
                <button
                    id={id}
                    ref={buttonRef}
                    {...elem('mainTextButton', { ...props, muted: !selectionPlaceholder })}
                    onFocus={handleOpenModal}
                    onBlur={handleButtonBlur}
                    onKeyUp={handleButtonKeyPress}
                >
                    {selectionPlaceholder || inputPlaceholder}
                </button>
            </FieldWrapper>
            <Modal
                {...elem('modal', props)}
                isOpen={isOpen}
                ariaHideApp={false}
                onRequestClose={handleCloseModal}
                contentLabel={modalContentLabel}
            >
                <LocationSelectorDialogWithGoogleLoader
                    apiKey={apiKey}
                    language={language}
                    region={region}
                    {...additionalGoogleProps}
                    inputPlaceholder={inputPlaceholder}
                    hasRadius={hasRadius}
                    minRadius={minRadius}
                    maxRadius={maxRadius}
                    radiusStep={radiusStep}
                    renderRadiusLabel={renderRadiusLabel}
                    onRemoveLocation={onRemoveLocation}
                    doneLabel={doneLabel}
                    country={country}
                    withoutLocationCards={withoutLocationCards}
                    initialMapAddress={initialMapAddress}
                    placeTypes={placeTypes}
                    noSuggestionsPlaceholder={noSuggestionsPlaceholder}
                    showCountryInSuggestions={showCountryInSuggestions}
                    onLocationAutocompleteError={onLocationAutocompleteError}
                    onUpdateLocation={onUpdateLocation}
                    onRemoveAllLocations={onRemoveAllLocations}
                    selectedLocations={selectedLocations}
                    getMarkers={getMarkers}
                    onAddLocation={handleAddLocation}
                    onCloseModal={handleCloseModal}
                />
            </Modal>
        </div>
    );
};

LocationSelector.displayName = 'LocationSelector';

LocationSelector.defaultProps = {
    hasRadius: true,
    radiusDefaultValue: 1,
    minRadius: 1,
    maxRadius: 100,
    radiusStep: 1,
    withoutLocationCards: false,
    initialMapAddress: undefined,
    selectedLocations: [],
    showCountryInSuggestions: true,
    onBlur: () => null,
    additionalGoogleProps: {},
    region: undefined,
    placeTypes: ['(regions)'],
    onLocationAutocompleteError: () => null,
};
