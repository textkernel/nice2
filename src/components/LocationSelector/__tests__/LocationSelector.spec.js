import React from 'react';
import LocationSelector from '../LocationSelector';

describe('LocationSelector component', () => {
    const selectedLocations = [];

    const onAddLocationMock = jest.fn();
    const onUpdateLocationMock = jest.fn();
    const onRemoveLocationMock = jest.fn();
    const onRemoveAllLocationsMock = jest.fn();

    function shallowLocationSelector() {
        return shallow(
            <LocationSelector
                apiKey="abc"
                selectedLocations={selectedLocations}
                country="NL"
                language="EN"
                radiusUnits="km"
                radiusUnitDisplayText="42"
                minRadius={1}
                maxRadius={100}
                radiusStep={1}
                autocompletePlaceholder="autocompletePlaceholder"
                noSuggestionsPlaceholder="noSuggestionsPlaceholder"
                selectionPlaceholder="selectionPlaceholder"
                mainPlaceholder="mainPlaceholder"
                doneLabel="doneLabel"
                clearLabel="clearLabel"
                contentLabel="contentLabel"
                onAddLocation={onAddLocationMock}
                onUpdateLocation={onUpdateLocationMock}
                onRemoveLocation={onRemoveLocationMock}
                onRemoveAllLocations={onRemoveAllLocationsMock}
            />
        );
    }

    it('should render LocationSelector properly', () => {
        const wrapper = shallowLocationSelector();
        expect(wrapper).toMatchSnapshot();
    });

    it.skip('should open modal window by clicking on the Text component', () => {
        const wrapper = shallowLocationSelector();
        expect(wrapper.find('.LocationSelector__autocomplete')).toHaveLength(0);
        wrapper.find('LocationSelector__mainTextInput').simulate('click');
        expect(wrapper.find('.LocationSelector__autocomplete')).toHaveLength(1);
    });
});
