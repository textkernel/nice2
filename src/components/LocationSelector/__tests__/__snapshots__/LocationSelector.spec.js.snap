// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LocationSelector component should render LocationSelector properly 1`] = `ShallowWrapper {}`;
